### ISSUE TITLE
DWARF section lacks clarifications about CFI 

### ISSUE THREAD

**Issue by serhiihuralniksc** [riscv-non-isa/riscv-elf-psabi-doc#239]
DWARF CFI is an encoding for unwind information, see ["6.4.1 Structure of Call Frame Information"](https://dwarfstd.org/doc/DWARF4.pdf).

Current psABI version does not mention initial register rules that must be used by the unwinding agent before processing FDE unwind instructions. Though CIE program is recommended as an initializer:

> 1. Initialize a register set by reading the initial_instructions field of the associated CIE.

nevertheless at least some implementations do not emit a full initialization in CIE and only define CFA. E.g.:
```
$ cat foo.c

extern int bar(int);

int foo(int a) {
    return bar(a + 10) - 100;
}

$ clang -target riscv64 foo.c -c -funwind-tables -g
$ llvm-dwarfdump foo.o -debug-frame
foo.o:	file format ELF64-riscv

.debug_frame contents:


.eh_frame contents:

00000000 00000010 ffffffff CIE
  Version:               1
  Augmentation:          "zR"
  Code alignment factor: 1
  Data alignment factor: -8
  Return address column: 1
  Augmentation data:     1B

  DW_CFA_def_cfa: reg2 +0

00000014 00000020 00000018 FDE cie=00000018 pc=00000000...00000000
  DW_CFA_advance_loc: 2
  DW_CFA_def_cfa_offset: +32
  DW_CFA_advance_loc: 4
  DW_CFA_offset: reg1 -8
  DW_CFA_offset: reg8 -16
  DW_CFA_advance_loc: 2
  DW_CFA_def_cfa: reg8 +0
  DW_CFA_nop:
  DW_CFA_nop:
  DW_CFA_nop:
  DW_CFA_nop:
  DW_CFA_nop:
  DW_CFA_nop:
  DW_CFA_nop:

```
In this regard an unwinding agent must assume some rules for other registers before processing FDE instructions.

And here we have an inconsistency between the standard and existing implementations - there is nothing to fill this gap for other registers. There are at least two options to resolve this:
- psABI declares an initial rules for registers that must be applied by the unwinder even before CIE instructions are processed. This lets us to avoid unwinder states with unknown register rules and at the same time allows to save a size of objects due to implicit rules. I can imagine that callee-saved regs could get a `same value` rule, some other - `undefined`, so on.
- require each implementation to emit a full initialization for each CIE, this is the least favored I guess.

Likely some other choices are possible. Anyways an actual decision should be reflected in the psABI.

**Comment by jrtc27** [comment]
Basically you just want us to copy aadwarf64 §4.3? It's a shame each psABI has to define that rather than it just being a generic thing.

**Comment by serhiihuralniksc** [comment]
@jrtc27 Indeed, this is what I meant. Something should be likely revised to reflect RISC-V specifics, but the overall idea is the same.
>  It's a shame each psABI has to define that rather than it just being a generic thing.

They have CIE sequence for this purpose. Maybe it was effective enough at the times CFI section first appeared in that standard, who knows.
